src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -0,0 +1,289 @@
+            BitbucketDeployment deployment = descriptor()
+                    .getBitbucketDeploymentFactory()
+                    .createDeployment(run, environment);
+
+            BitbucketSCMRepository bitbucketSCMRepo = revisionAction.getBitbucketSCMRepo();
+            String revisionSha = revisionAction.getRevisionSha1();
+            descriptor().getDeploymentPoster().postDeployment(bitbucketSCMRepo.getServerId(),
+                    bitbucketSCMRepo.getProjectKey(), bitbucketSCMRepo.getRepositorySlug(), revisionSha, deployment,
+                    run, listener);
+        } catch (RuntimeException e) {
+            // This shouldn't happen because deploymentPoster.postDeployment doesn't throw anything. But just in case,
+            // we don't want to throw anything and potentially stop other steps from being executed
+            String errorMsg =
+                    format("An error occurred when trying to post the deployment to Bitbucket Server: %s", e.getMessage());
+            listener.error(errorMsg);
+            LOGGER.info(errorMsg);
+            LOGGER.log(Level.FINE, "Stacktrace from deployment post failure", e);
+        }
+    }
+
+    private BitbucketDeploymentEnvironment getEnvironment(Run<?, ?> run, TaskListener listener) {
+        return new BitbucketDeploymentEnvironment(environmentKey,
+                getOrGenerateEnvironmentName(environmentName, run, listener),
+                environmentType,
+                getEnvironmentUri(environmentUrl, listener));
+    }
+
+    @CheckForNull
+    private BitbucketDeploymentEnvironmentType getEnvironmentType(@CheckForNull String environmentType) {
+        if (isBlank(environmentType)) {
+            return null;
+        }
+        return BitbucketDeploymentEnvironmentType.fromName(environmentType)
+                .orElseGet(() -> {
+                    LOGGER.warning(format("DeployedToEnvironmentNotifierStep: Invalid environment type '%s'. Saving step without environment type.", environmentType));
+                    return null;
+                });
+    }
+
+    @CheckForNull
+    private URI getEnvironmentUri(String environmentUrl, TaskListener listener) {
+        if (isBlank(environmentUrl)) {
+            return null;
+        }
+        try {
+            return new URI(environmentUrl);
+        } catch (URISyntaxException x) {
+            listener.getLogger().println(format("DeployedToEnvironmentNotifierStep: Invalid environment URL '%s'. Posting deployment without a URL instead.", this.environmentUrl));
+            return null;
+        }
+    }
+
+    private String getOrGenerateEnvironmentKey(@CheckForNull String environmentKey) {
+        if (!isBlank(environmentKey)) {
+            return environmentKey;
+        }
+        return UUID.randomUUID().toString();
+    }
+
+    private String getOrGenerateEnvironmentName(@CheckForNull String environmentName, Run<?, ?> run,
```

Comment:
As I mentioned in my [reply to Daniel's comment](https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265#discussion_r702551189), the generating the environment name is a little different to other methods we call in the constructor because it uses information from the `run`.
`getEnvironment` does not call to `getOrGenerateEnvironmentName`, so I'm not sure what 'each time' is that you're referring to

Reply:
I thought `getEnvironment` does. call to `getOrGenerateEnvironmentName` at [line 148](https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/265/files/c664303044295b48c85fddc416cce8770205c5f6#diff-5aa1fc6c9c93213f91766e67821cdd817799b361038f930454aaca7c1dda9b5cR148). But maybe I'm understanding things wrong, so long as it's something we've considered